compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -933,6 +934,27 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         }
     }
 
+    fn report_ambiguous_type_parameter(&self, err: &mut Diagnostic, arg: GenericArg<'tcx>) {
+        if let GenericArgKind::Type(ty) = arg.unpack()
+            && let ty::Infer(ty::TyVar(ty_vid)) = *ty.kind()
+        {
+            let mut inner = self.inner.borrow_mut();
+            let ty_vars = &inner.type_variables();
+            let var_origin = ty_vars.var_origin(ty_vid);
+            if let TypeVariableOriginKind::TypeParameterDefinition(name, Some(def_id)) =
+                var_origin.kind
+                && let Some(parent_def_id) = self.tcx.parent(def_id).as_local()
+                && let Some(hir::Node::Item(item)) = self.tcx.hir().find_by_def_id(parent_def_id)
+                && let hir::ItemKind::Impl(impl_item) = item.kind
+                && let Some(trait_ref) = &impl_item.of_trait
+                && let Some(did) = trait_ref.trait_def_id()
+                && self.tcx.generics_of(did).params.iter().any(|param| param.name == name)
```

Comment:
Do we need these checks? Will the diagnostic be too noisy if we just always report the `help` on the `def_id`?

Reply:
While for `Option` it is a bit redundant, I'd hope that for user defined types this will be helpful.

Maybe not emitting it on type declarations, but on any function, method or impl could remove those weird cases?

Reply:
Ok. I'm going to fix it.